docs/design/data_safety.md

```diff
@@ -0,0 +1,153 @@
+    // Invariant 2: there are 7 elements in the vector
+    weekday_names: Vec<String>,
+}
+
+impl WeekdayNamesBad {
+    pub fn get_first_weekday_name(&self) -> &str {
+        &self.weekday_names[0]
+    }
+    pub fn get_weekday_name_at_index(&self, idx: usize) -> Option<&str> {
+        if idx >= 7 {
+            return None;
+        }
+        Some(&self.weekday_names[idx])
+    }
+}
+```
+
+This struct has 2 invariants, and each of the functions could panic if those invariants are not upheld.
+
+Below are 3 approaches to resolve this issue that are consistent with the ICU4X style.
+
+*Editor's note: the examples in this document are tested along with the tutorials.*
+
+### Solution 1: Garbage In, Garbage Out
+
+Change the functions to return default fallback values if the data is not in the expected form.
+
+```rust
+#[derive(serde::Deserialize)]
+pub struct WeekdayNamesGIGO {
+    // WEAK Invariant 1: the first element in the vector is the default value
+    // WEAK Invariant 2: there are 7 elements in the vector
+    weekday_names: Vec<String>,
+}
+
+impl WeekdayNamesGIGO {
+    pub fn get_first_weekday_name(&self) -> &str {
+        match self.weekday_names.get(0) {
+            Some(v) => v,
+            None => {
+                debug_assert!(false, "weekday_names is empty");
+                ""
+            }
+        }
+    }
+    pub fn get_weekday_name_at_index(&self, idx: usize) -> Option<&str> {
+        self.weekday_names.get(idx).map(String::as_str)
+    }
+}
+```
+
+### Solution 2: Remove Internal Invariants
+
+Change the struct so that it doesn't have internal invariants.
+
+```rust
+#[derive(serde::Deserialize)]
+pub struct WeekdayNamesNoInvariants {
+    pub first_weekday_name: String,
+    pub remaining_weekdays: Vec<String>,
+}
+
+impl WeekdayNamesNoInvariants {
+    pub fn get_first_weekday_name(&self) -> &str {
+        &self.first_weekday_name
+    }
+    pub fn get_weekday_name_at_index(&self, idx: usize) -> Option<&str> {
+        if idx == 0 {
+            Some(&self.first_weekday_name)
+        } else {
+            self.remaining_weekdays.get(idx - 1).map(String::as_str)
+        }
+    }
+}
+```
+
+This struct has no internal invariants. The deserialization step automatically enforces that the default value is present.
+
+### Solution 3: Use Helper Struct
+
+With this solution, check that there are the expected number of items at deserialization time. Leverage Serde's [try_from](https://serde.rs/container-attrs.html#try_from) attribute to perform the validation.
+
+```rust
+use icu_provider::DataError;
+
+#[derive(serde::Deserialize)]
+struct WeekdayNamesInner {
+    pub weekday_names: Vec<String>,
+}
+
+#[derive(serde::Deserialize)]
+#[serde(try_from = "WeekdayNamesInner")]
+pub struct WeekdayNamesCheckedInvariants {
+    // Invariant: the vector is non-empty.
+    weekday_names: Vec<String>,
+}
+
+impl TryFrom<WeekdayNamesInner> for WeekdayNamesCheckedInvariants {
+    type Error = DataError;
```

Comment:
Note: I realized that this error gets promptly cast to `serde::de::Error::custom(...)` as part of serde::try_from

https://github.com/serde-rs/serde/blob/fb2fe409c8f7ad6c95e3096e5e9ede865c8cfb49/serde_derive/src/de.rs#L401

What would be the best practice error type for this position?

Reply:
Yeah we should ideally use serde's error type directly

Reply:
I think the most efficient type to use is `&'static str` because serde only exposes a trait, and the trait casts whatever comes in as `Display`.

Reply:
yeah let's do that